SimTracker/SiStripDigitizer/plugins/SiStripDigitizerAlgorithm.cc

```diff
@@ -305,6 +430,79 @@ void SiStripDigitizerAlgorithm::digitize(edm::DetSet<SiStripDigi>& outdigi,
     }
   }
 
+  if (includeAPVSimulation_) {
+    // Get index in apv baseline distributions corresponding to z of detSet and PU
+    const StripTopology* topol = dynamic_cast<const StripTopology*>(&(det->specificTopology()));
+    LocalPoint localPos = topol->localPosition(0);
+    GlobalPoint globalPos = det->surface().toGlobal(Local3DPoint(localPos.x(), localPos.y(), localPos.z()));
+    float detSet_z = fabs(globalPos.z());
+    std::vector<double>::iterator high =
+        std::upper_bound(apvBaselines_zBinEdges_.begin(), apvBaselines_zBinEdges_.end(), detSet_z);
+    unsigned int detSet_zBin = high - apvBaselines_zBinEdges_.begin() - 1;
+    high = std::upper_bound(apvBaselines_puBinEdges_.begin(), apvBaselines_puBinEdges_.end(), nTruePU_);
+    unsigned int puBin = high - apvBaselines_puBinEdges_.begin() - 1;
+
+    // Get the corresponding APV baseline distribution for this subdetector and layer
+    TH1F* apvBaselineDistribution = nullptr;
+    int layer = -1;
+    if (SubDet == 3) {
+      layer = tTopo->tibLayer(detId);
+      apvBaselineDistribution = &apvBaselineHistograms_tib_[layer - 1][detSet_zBin][puBin];
+    } else if (SubDet == 5) {
+      layer = tTopo->tobLayer(detId);
+      apvBaselineDistribution = &apvBaselineHistograms_tob_[layer - 1][detSet_zBin][puBin];
+    }
+
+    // Store SCD, before APV sim
+    for (int strip = 0; strip < numStrips; ++strip) {
+      outStripAmplitudes.push_back(SiStripRawDigi(detAmpl[strip] / theElectronPerADC));
+      ;
+    }
+
+    // Simulate APV response for each strip
+    if (SubDet == 3 || SubDet == 5) {
+      for (int strip = 0; strip < numStrips; ++strip) {
+        if (detAmpl[strip] > 0) {
+          // Convert charge from electrons to fC
+          double stripCharge = detAmpl[strip] * apv_fCPerElectron;
+
+          // Get APV baseline
+          double baselineV = apvBaselineDistribution->GetRandom();
```

Comment:
I'm not sure, but perhaps something similar to this may be useful:

https://github.com/cms-sw/cmssw/blob/07d3399c66e34a7aa14773466b70643db096a7f7/Mixing/Base/src/PileUp.cc#L49-L63

Or something like std::discrete_distribution could be used, but I don't know if this could use CLHEP::HepRandomEngine.

Reply:
I suppose that would work. By quick look `CLHEP::HepRandomEngine` does not satisfy [*UniformRandomBitGenerator*](https://en.cppreference.com/w/cpp/named_req/UniformRandomBitGenerator) requirement out of the box (needed for `std::discrete_distribution::operator()()`), but a thin wrapper might be possible.

(personally I'd really like if TH1 could be avoided as it has member data unnecessary for this case, leading to memory bloat, and it has caused thread-safety problems earlier within MixingModule)

Reply:
Actually, CLHEP::RandGeneral performs the same task, so I expect this would work without an additional wrapper.  I missed this first time looking at through CLHEP.